schema.ts

```diff
@@ -1232,6 +1232,82 @@ export interface ObjectSchemaDefinition<K extends string, L extends string>
   // TODO(dweitzman): Only support options in the typing when the codaType is ValueHintType.SelectList.
 }
 
+/**
+ * Discriminent used to determine the permission type
+ * @hidden
+ */
+export enum PrincipalType {
+  User = 'user',
+  Group = 'group',
+  Domain = 'domain',
+  Public = 'public',
+}
+
+/**
+ * This type definition represents properties required for all permissions.
+ *
+ * TODO(sam): Unhide this
+ * @hidden
+ */
+interface BasePermission {
+  type: PrincipalType;
+  rowId: string | number;
+  // TODO(sam): Add options describing inheritance of permissions
+}
+
+/**
+ * This grants access to a specific user by their user ID.
+ *
+ * TODO(sam): Unhide this
+ * @hidden
+ */
+interface UserPermission extends BasePermission {
+  type: PrincipalType.User;
+  userId: string | number;
+}
+
+/**
+ * This grants access to a specific group by its group ID.
+ *
+ * TODO(sam): Unhide this
+ * @hidden
+ */
+interface GroupPermission extends BasePermission {
+  type: PrincipalType.Group;
+  groupId: string | number;
+}
+
+/**
+ * This grants access to users within a specific domain.
+ *
+ * TODO(sam): Unhide this
+ * @hidden
+ */
+interface DomainPermission extends BasePermission {
+  type: PrincipalType.Domain;
+  domainName: string;
+}
+
+/**
+ * This grants access to any user in the world.
+ *
+ * Generally this would apply to an entity where anyone with access to the url can view the item
+ *
+ * TODO(sam): Unhide this
+ * @hidden
+ */
+interface PublicPermission extends BasePermission {
```

Comment:
I see in our public API we call this an "Anyone" principal, not sure if we want to be consistent with that, @ekoleda-codaio any opinions?

Reply:
Happy to be consistent here with the naming